docs: improve httpbin charm by removing defer() and adding collect_status

dwilding opened the pull request:
This PR makes a few improvements to the [httpbin demo charm](https://github.com/canonical/operator/tree/main/examples/httpbin-demo).

Main improvements:
- Added a class attribute for the workload container
- Made sure that log-level is always validated before being put into a Pebble layer
- Removed `defer()` and instead retry the container calls a few times
- Use collect_status to report the unit status in a consolidated way

Drive-bys:
- Added constants `CONTAINER_NAME` and `SERVICE_NAME` to help differentiate between different uses of `httpbin`
- Added a unit test for the case where an uppercase log level is configured
- Made the formatting of comments consistent
- As in #1734, specified a minimum Ops version and adjusted the formatting of tox.ini

james-garner-canonical commented on examples/httpbin-demo/src/charm.py:
```diff
@@ -32,43 +31,44 @@ class HttpbinDemoCharm(ops.CharmBase):
+        framework.observe(self.on.collect_unit_status, self._on_collect_status)
         framework.observe(self.on['httpbin'].pebble_ready, self._on_httpbin_pebble_ready)
         framework.observe(self.on.config_changed, self._on_config_changed)
+        self.log_level = str(self.config['log-level'])  # str() is here for the type checker.
```
I wonder if the `.lower()` call go here instead? What's actually allowed for the log level? e.g. is `DeBuG` ok?

tonyandrewmeyer replied:
More radical idea:

```python
self.config = self.load_config(HttpBinConfig, errors='blocked')
```

With a class like:

```python
@dataclasses.dataclass(frozen=True, kw_only=True)
class HttpBinConfig:
    log_level: str = 'info'
    """Configures the log level of gunicorn.

    Acceptable values are: "debug", "info", "warning", "error" and "critical"
    """

    def __post_init__(self):
        log_level = self.log_level.lower()
        if log_level not in VALID_LOG_LEVELS:  # That could probably be a class attribute.
            raise ValueError(f'`log-level` must be one of {", ".join(VALID_LOG_LEVELS)}, not {log_level}')
        object.__setattr__(self, 'log_level', log_level)
```

And then it's accessed as `self.config.log_level` when needed.

james-garner-canonical replied:
I'm in favour of using `load_config`, but I think that isn't possible here till the next release (which we could wait for), or we update the CI to pull `ops` from `main` when running the example charm tests -- or I guess we could just update the `pyproject.toml` for this charm to specify `main`?

tonyandrewmeyer replied:
Yeah, I was assuming a "wait and require 2.23", but fair point that it's still ~2 weeks away and @dwilding (presumably) doesn't want to add a blocked PR.

So I guess my suggestion would be we do this but in a follow-up PR, but as part of that, don't take too much care to improve the config handling because we know an improvement is coming.

(I'm happy to open one for using it at the end of the month if @dwilding doesn't want the extra work).

james-garner-canonical replied:
How about if @dwilding is on board with using `load_config`, we temporarily change this charm's `requirements.txt` to specify `main`?
```
ops @ git+https://github.com/canonical/operator@main
```
 That way the post-release follow-up PR would only need to change that.

dwilding replied:
I'm not sure how I feel about this. I agree with switching to `load_config`, in the spirit of keeping our example charms fresh on the latest features. I'm less sure about requiring Ops from main, even temporarily, as we wouldn't want someone to think that we recommend doing it for a real charm. I'll work on the other updates and see how I feel after that!

dwilding replied:
I prefer to leave it as-is for now, so that we don't need to require Ops from main. I've created #1839 for one of us to work on later

james-garner-canonical commented on examples/httpbin-demo/src/charm.py:
```diff
@@ -32,43 +31,44 @@ class HttpbinDemoCharm(ops.CharmBase):
+        """Report the status of the workload (runs after each event)."""
+        if self.log_level.lower() not in VALID_LOG_LEVELS:
+            event.add_status(ops.BlockedStatus(f"invalid log level: '{self.log_level}'"))
+            return
+        if not self.container.can_connect():
+            event.add_status(ops.MaintenanceStatus('waiting for container'))
+            return
+        try:
+            if not self.container.get_service('httpbin').is_running():
+                event.add_status(ops.MaintenanceStatus('waiting for workload'))
+                return
+        except ops.ModelError:
+            event.add_status(ops.MaintenanceStatus('waiting for Pebble'))
+            return
```
Is this better without  the early `return`s, since the idea is that collect status will prioritise the statuses added with `add_status`?

tonyandrewmeyer replied:
Yes, all the returns in collect-status should be removed - the idea is that you add all the relevant statuses and let ops pick which one to use.

james-garner-canonical commented on examples/httpbin-demo/src/charm.py:
```diff
@@ -32,43 +31,44 @@ class HttpbinDemoCharm(ops.CharmBase):
+            event.add_status(ops.MaintenanceStatus('waiting for container'))
+            return
+        try:
+            if not self.container.get_service('httpbin').is_running():
+                event.add_status(ops.MaintenanceStatus('waiting for workload'))
+                return
+        except ops.ModelError:
```
Why do we get a `ModelError` here?

dwilding replied:
In [the doc](https://ops.readthedocs.io/en/latest/reference/ops.html#ops.Container.get_service) it says that `get_service()` raises `ModelError` if the service is not found, which I'm interpreting as Pebble not being fully ready (although `can_connect()` has already returned True, so Pebble must be somewhat ready - modulo Tony's point about race conditions). Would it be better to take a different approach?

james-garner-canonical commented on examples/httpbin-demo/src/charm.py:
```diff
@@ -32,43 +31,44 @@ class HttpbinDemoCharm(ops.CharmBase):
-        else:
-            # In this case, the config option is bad, so block the charm and notify the operator.
-            self.unit.status = ops.BlockedStatus(f"invalid log level: '{log_level}'")
+        if self.log_level.lower() not in VALID_LOG_LEVELS or not self.container.can_connect():
```
I suspect so, but I'll let @tonyandrewmeyer make the call here. If so, would we want to use `tenacity`, or just use a loop?

tonyandrewmeyer replied:
I really like `tenacity` and I've used it a lot in the past. However, it's a big library to include if you have simple needs. So I would generally avoid using it in charms.

I don't love `can_connect()` because it introduces a (extraordinarily unlikely) race condition: Pebble might be able to connect at the time of the `can_connect()` call and then fail in the following call (a millisecond or two later).

My preference is for a pattern like:

```python
for attempt in range(5):
    try:  # the linter will complain about try/except in a loop, so will need a noqa
        self.container.add_layer(...)
        self.container.replan()
    except (ops.pebble.APIError, ops.pebble.ConnectionError):
        logger.info(...)
        time.sleep(2**attempt)  # or whatever formula you like that will give reasonable times
    else:
        break
else:
    logger.warning(...)
    # We expect that there will be a pebble-ready event in the future that will handle this.
    return
```

dwilding replied:
Resolved per Tony's suggestion:

```python
# Update the configuration of the workload.
# We might not be able to access the workload container yet, so we'll try a few times.
max_attepts = 3
for attempt in range(max_attepts):
    try:
        self.container.add_layer('httpbin', self._pebble_layer, combine=True)
        self.container.replan()
    except (ops.pebble.APIError, ops.pebble.ConnectionError):  # noqa: PERF203 (try-except in loop)
        logger.info('Unable to reconfigure gunicorn (attempt %d)', attempt + 1)
        time.sleep(2**attempt)
    else:
        break
else:
    logger.warning('Unable to reconfigure gunicorn after %d attempts.', max_attepts)
    # We expect that there'll be a pebble-ready event in the future,
    # which will configure and start the workload.
    return
```

tonyandrewmeyer commented on examples/httpbin-demo/src/charm.py:
```diff
@@ -32,43 +31,44 @@ class HttpbinDemoCharm(ops.CharmBase):
+
+    def _on_collect_status(self, event: ops.CollectStatusEvent):
+        """Report the status of the workload (runs after each event)."""
+        if self.log_level.lower() not in VALID_LOG_LEVELS:
+            event.add_status(ops.BlockedStatus(f"invalid log level: '{self.log_level}'"))
+            return
```
If we go with my more radical suggestion, this wouldn't be needed. Or if we think it would be better we could remove the `errors=`, catch the `ValueError` in `__init__` and then also have here something like:

```python
try:
    self.load_config(HttpBinConfig)
except ValueError as e:
    event.add_status(ops.BlockedStatus(str(e)))
```

tonyandrewmeyer commented on examples/httpbin-demo/src/charm.py:
```diff
@@ -32,43 +31,44 @@ class HttpbinDemoCharm(ops.CharmBase):
-        container.add_layer('httpbin', self._pebble_layer, combine=True)
+        # Add initial Pebble config layer using the Pebble API.
+        self.container.add_layer('httpbin', self._pebble_layer, combine=True)
         # Make Pebble reevaluate its plan, ensuring any services are started if enabled.
-        container.replan()
-        self.unit.status = ops.ActiveStatus()
+        self.container.replan()
```
Technically, you can get a pebble-ready and then fail to talk to Pebble (the container was up and then by the time you got the event it was down, or network issues, or whatever). Like with the `can_connect()` race condition, this should be extremely unlikely.

Personally, I think given how unlikely it is, it's reasonable to have the code like you have it. If it fails, the unit goes to error state and someone investigates (which they probably should, because your cloud probably has issues if this is happening) or Juju automatically retries if that setting is enabled. Some charmers strongly dislike error status and try to avoid it at all costs, but I feel like this is a legitimate error status.

But I would probably add a comment saying something like this. For example, "If we cannot connect to Pebble, let the unit go into error status, where this hook will be retried or the Juju user will investigate".

dwilding replied:
Resolved by adding this comment:

```
# In rare cases, these calls could fail because the workload container became unavailable.
# If this happens, we'll let the unit go into error status. The hook will be retried, or
# the Juju user can investigate.
```

tonyandrewmeyer commented on examples/httpbin-demo/src/charm.py:
```diff
@@ -32,43 +31,44 @@ class HttpbinDemoCharm(ops.CharmBase):
+                event.add_status(ops.MaintenanceStatus('waiting for workload'))
+                return
+        except ops.ModelError:
+            event.add_status(ops.MaintenanceStatus('waiting for Pebble'))
```
Nit: it feels like the Juju user probably shouldn't need to know a lot about Pebble (although once they start to debug they certainly would) so maybe "waiting for the workload container" would be more the terms they'd expect?

tonyandrewmeyer commented on examples/httpbin-demo/tests/unit/test_charm.py:
```diff
@@ -86,6 +86,23 @@ def test_config_changed_valid_cannot_connect():
+    # Assert:
+    updated_plan = state_out.get_container(container.name).plan
+    gunicorn_args = updated_plan.services['httpbin'].environment['GUNICORN_CMD_ARGS']
+    assert gunicorn_args == '--log-level DEBUG'
```
I'm confused by this one (and haven't run the tests locally, sorry). Doesn't the log level get `lower()`d? It seems like this should be `debug` not `DEBUG`.

tonyandrewmeyer reviewed (commented):
Thanks for doing this! A few suggestions, including one you should feel free to reject to use the new config class functionality.

dwilding commented on examples/httpbin-demo/src/charm.py:
```diff
@@ -32,43 +32,55 @@ class HttpbinDemoCharm(ops.CharmBase):
+        """Report the status of the workload (runs after each event)."""
+        if self.log_level.lower() not in VALID_LOG_LEVELS:
+            event.add_status(ops.BlockedStatus(f"invalid log level: '{self.log_level}'"))
+        try:
+            if not self.container.get_service('httpbin').is_running():
+                event.add_status(ops.MaintenanceStatus('waiting for workload'))
+        except (ops.pebble.ConnectionError, ops.ModelError):
+            event.add_status(ops.MaintenanceStatus('waiting for workload container'))
```
@james-garner-canonical & @tonyandrewmeyer would you mind taking a look at this updated logic? I think we need to catch `ops.ModelError` in case the service is not found. I don't think we need to catch `ops.pebble.APIError`, as `get_service` only asks the Pebble API for a list of services, which shouldn't return an error. What do you think?

dwilding replied:
For `ModelError`, I was going off the doc: https://ops.readthedocs.io/en/latest/reference/ops.html#ops.Container.get_service

james-garner-canonical replied:
Or should we not catch `ModelError` and let it take us to error status instead, since it's an error if we got the service name wrong?

dwilding replied:
To James's question - that's one thing I'm unclear about. Is a service name mismatch between charm code and layer definition the only reason we'd get `ModelError`? Is it possible for, e.g., Pebble to send over an empty list of services before it's fully ready?

dwilding replied:
I thought about this more and decided to go with a variation on James's suggestion. For the "unexpected" errors, `pebble.APIError` and `ModelError`, catch the error then log the error and re-raise it.

```python

try:
    if not self.container.get_service(SERVICE_NAME).is_running():
        # We can connect to Pebble in the container, but the service isn't running.
        event.add_status(ops.MaintenanceStatus('waiting for workload'))
except ops.pebble.ConnectionError:
    # We can't connect to Pebble in the container.
    event.add_status(ops.MaintenanceStatus('waiting for workload container'))
except ops.pebble.APIError:
    # It's technically possible (but unlikely) for Pebble to have an internal error.
    logger.error('Unable to fetch service info from Pebble')
    raise
except ops.ModelError:
    # We asked for the status of a service that doesn't exist in Pebble.
    # If this error happens, it means there's a bug in the charm.
    logger.error("Unable to find service '%s' in service info", SERVICE_NAME)
    raise
```

It's possible for the collect-status handler to run before the charm has added a layer to Pebble (e.g., in the case of an invalid config value). So I've updated the unit tests to provide a mock layer to the container, which ensures that `get_service()` can always find the service.

dwilding replied:
Please ignore my previous comment. I realised that the approach doesn't work. `ModelError` will be raised as part of normal operation because the workload doesn't have a default Pebble layer (it's not a rock). In other words, anytime the collect-status handler runs before the charm has added a layer, `get_service()` will be able to connect to Pebble, but there won't be any services yet, so we'll get `ModelError`.

I've changed the logic to this:

```python
try:
    if not self.container.get_service(SERVICE_NAME).is_running():
        # We can connect to Pebble in the container, but the service hasn't started yet.
        event.add_status(ops.MaintenanceStatus('waiting for workload'))
except ops.ModelError:
    # We can connect to Pebble in the container, but the service doesn't exist. This is
    # most likely because we haven't added a layer yet.
    event.add_status(ops.MaintenanceStatus('waiting for workload container'))
except ops.pebble.ConnectionError:
    # We can't connect to Pebble in the container. This is most likely because the
    # container hasn't started yet.
    event.add_status(ops.MaintenanceStatus('waiting for workload container'))
except ops.pebble.APIError:
    # It's technically possible (but unlikely) for Pebble to have an internal error.
    logger.error('Unable to fetch service info from Pebble')
    raise
```

And I've reverted the mock layer in the unit tests.

dwilding commented on examples/httpbin-demo/requirements.txt:
```diff
@@ -1 +1 @@
-ops < 4
+ops>2,<4
```
Adding a minimum version (as in #1734) in anticipation of switching to `load_config` after the next release (#1839)

james-garner-canonical commented on examples/httpbin-demo/src/charm.py:
```diff
@@ -32,43 +32,55 @@ class HttpbinDemoCharm(ops.CharmBase):
-            # The config is good, so update the configuration of the workload
-            container = self.unit.get_container('httpbin')
-            # Push an updated layer with the new config
+        if self.log_level.lower() not in VALID_LOG_LEVELS:
```
```suggestion
        if self.log_level not in VALID_LOG_LEVELS:
```

dwilding replied:
Oops, thanks

james-garner-canonical commented on examples/httpbin-demo/src/charm.py:
```diff
@@ -32,43 +32,55 @@ class HttpbinDemoCharm(ops.CharmBase):
+            return
+        # Update the configuration of the workload.
+        # We might not be able to access the workload container yet, so we'll try a few times.
+        max_attempts = 3
+        for attempt in range(max_attempts):
```
Very nitpicky and possibly a matter of taste, but I'd inline `max_attempts`
```suggestion
        for attempt in range(3):
```
Oh unless you want to add it to the logging like 'attempt 1 of 3' kind of thing.

dwilding replied:
I'm using `max_attempts` in the logging message if all attempts fail:

```python
logger.warning('Unable to reconfigure gunicorn after %d attempts.', max_attempts)
```

So I'll leave it as-is, unless you feel that this isn't necessary

james-garner-canonical replied:
Sounds good, I didn't notice that

tonyandrewmeyer commented on examples/httpbin-demo/src/charm.py:
```diff
@@ -17,58 +17,83 @@
+        framework.observe(self.on.collect_unit_status, self._on_collect_status)
+        framework.observe(self.on[CONTAINER_NAME].pebble_ready, self._on_httpbin_pebble_ready)
         framework.observe(self.on.config_changed, self._on_config_changed)
+        self.log_level = str(self.config['log-level']).lower()  # str() is for the type checker.
```
Personally, I like charms doing as little as possible in `__init__`, but the Charm Tech position (from the discussion in Frankfurt) is that they can do anything they like.

There's going to be a follow-up to this PR that uses `load_config` instead, so my preference is still to leave things as they are and then finalise config access in that PR.

dwilding commented:
Per discussion at Charm Tech standup, I've added log-level validation at the top of the pebble-ready handler. I've also removed `self.log_level`. Instead, the logic for loading and validating log-level is duplicated everywhere it's needed. This is in anticipation of @tonyandrewmeyer switching to [`load_config`](https://ops.readthedocs.io/en/latest/reference/ops.html#ops.CharmBase.load_config) after that's released this month.

One other adjustment: I changed `_pebble_layer` from a property to a normal method (now called `_make_pebble_layer`) so that it can take log-level as an argument.
